---
title: "Assignement"
author: "Paula Iller"
date: "2024-04-19"
output: html_document
---

```{r}

#BiocManager::install("ETHZ-INS/epiwraps") # update epiwrap 
suppressPackageStartupMessages({
  library(epiwraps)
  library(AnnotationHub)
  library(MotifDb)
  library(memes)
  library(universalmotif)
  library(ensembldb)
  library(ggplot2)
})
ah <- AnnotationHub(localHub=TRUE)
ensdb <- ah[["AH89211"]] # mouse ensembldb object
```
```{r}
## Download the data
download.file("https://ethz-ins.org/content/w7/atac.chr19.bam", "atac.chr19.bam")
download.file("https://ethz-ins.org/content/w7/atac.chr19.bam.bai", "atac.chr19.bam.bai")

# note that we could also do something like this programmatically:
tracks <- list.files(pattern="bw$")
names(tracks) <- gsub("\\.bw","",basename(tracks))
bam <- "atac.chr19.bam"
nf <- getNormFactors(tracks, useSeqLevels="19", nwind=5000L)


# create a track using all fragments, computing the coverage at each position (and force Ensembl naming)
bam2bw(bam, output_bw = "full_cov.bw", paired=TRUE, binWidth=10L, forceSeqlevelsStyle = "Ensembl") 

# create a track using all fragments, but computing the number of cuts/insertion sites at each position
epiwraps::bam2bw(bam, output_bw = "full_cuts.bw", paired=TRUE, binWidth=1L, type="ends", 
                 shift=c(4L,-5L), forceSeqlevelsStyle = "Ensembl")

# create a track using only nucleosome-free fragments, the number of cuts/insertion sites at each position 
bam2bw(bam, output_bw = "NF_cuts.bw", paired=TRUE, binWidth=1L, type="ends", minFragLength=30, 
       maxFragLength=120, shift=c(4L,-5L), forceSeqlevelsStyle = "Ensembl")

# create a track using only mono-nucleosome fragments, computing coverage
bam2bw(bam, output_bw = "mono_cov.bw", paired=TRUE, binWidth=10L, minFragLength=140,
       maxFragLength=220, forceSeqlevelsStyle = "Ensembl")

# create a track using only mono-nucleosome fragments, computing centers 
bam2bw(bam, output_bw = "mono_centers.bw", paired=TRUE, binWidth=5L, minFragLength=140,
       maxFragLength=220, type="center", forceSeqlevelsStyle = "Ensembl")

# create a track using only mono-nucleosome fragments, computing the number of cuts/insertion sites at each position 
bam2bw(bam, output_bw = "mono_cuts.bw", paired=TRUE, binWidth=1L, type="ends", minFragLength=140, 
       maxFragLength=220, shift=c(4L,-5L), forceSeqlevelsStyle = "Ensembl")
```

```{r, fig.width=8, fig.height=4}
# get KLF4 motif
motif_KLF4 <- MotifDb::query(MotifDb, c("KLF4","Mus"))[[1]]
motif2_KLF4 <- convert_motifs(motif_KLF4, class="TFBSTools-PFMatrix")

genome <- ah[["AH68356"]]
# get the sequence for chr19:
chr19 <- import(genome)["19"]

# find motif matches across chr19 using the converted motif
moi_KLF4 <- motifmatchr::matchMotifs(motif2_KLF4, chr19, out="positions", p.cutoff=1e-5)[[1]]
# convert to GRanges
moi_KLF4 <- as(setNames(moi_KLF4, names(chr19)), "GRanges")

# Use the converted motif here as well
motif_occ_KLF4 <- motifmatchr::matchMotifs(motif2_KLF4, chr19, out="positions", p.cutoff=1e-5)[[1]]
# convert to GRanges
motif_occ_KLF4 <- as(setNames(motif_occ_KLF4, names(chr19)), "GRanges")


#read signal of ATAC cuts around KLF4 motif
sm_KLF4 <- signal2Matrix(tracks, motif_occ_KLF4, w=5, extend=500)
#apply normalization factor
smb_KLF4 <- renormalizeSignalMatrices(sm_KLF4, scaleFactors = nf)

#plot heatmap of normalized regions
plotEnrichedHeatmaps(smb_KLF4, trim = 0.95, color=c("white", "darkred"), use_raster=FALSE)
```

```{r, fig.width=8, fig.height=4}
# get MAZ motif
motif_MAZ <- query(MotifDb, c("MAZ","Mus"))[[1]]
motif2_MAZ <- convert_motifs(motif_MAZ, class="TFBSTools-PFMatrix")

# find motif matches across chr19
moi_MAZ <- motifmatchr::matchMotifs(motif2_MAZ, chr19, out="positions", p.cutoff=1e-5)[[1]]
# convert to GRanges
moi_MAZ <- as(setNames(moi_MAZ,names(chr19)), "GRanges")

# find motif occurences/positions using matchMotifs
motif_occ_MAZ <- motifmatchr::matchMotifs(motif2_MAZ, chr19, out="positions", p.cutoff=1e-5)[[1]]
# convert to GRanges
motif_occ_MAZ <- as(setNames(motif_occ_MAZ, names(chr19)), "GRanges")

#read signal of ATAC cuts around MAZ motif
sm_MAZ <- signal2Matrix(tracks, motif_occ_MAZ, w=5, extend=500)
#apply normalization factor
smb_MAZ <- renormalizeSignalMatrices(sm_MAZ, scaleFactors = nf)

#plot heatmap of normalized regions
plotEnrichedHeatmaps(smb_MAZ, trim = 0.95, color=c("white", "darkred"), use_raster=FALSE)
```
